README.md
# Mosaic Image generator

- WIP & experimental
- implemented in Rust

```
Mosaic Image Generator

USAGE:
    mosion [FLAGS] [OPTIONS] <INPUT> <DIR> <OUTPUT>

FLAGS:
        --no_lab_cache    Disable caching of converted wall. Reduces memory usage, severely hurts performance
    -n                    No recursing in tile dir
    -y                    Allow overwriting output file
    -h, --help            Prints help information
    -V, --version         Prints version information

OPTIONS:
    -p <ACHUNKS>        Analysis concurrency (increases I/O utilization) (default=64)
    -s <CSCALE>         Compose Scale Factor (default=1)
    -m <MAPPER>         Mapping Algorithm (simple/alltoall) (default=alltoall)
    -k <TILEH>          Tile Height (default=64)
    -j <TILEW>          Tile Width (default=64)

ARGS:
    <INPUT>     Input "Wall" Image
    <DIR>       Tile Image Dir
    <OUTPUT>    Output Image
```

## Examples

```
mosion input.jpg tiles mosaic.png # generate mosaic.png using images in tiles/
mosion -s 2 input.jpg tiles mosaic.png # generate 2x size
```

## How does it work?

- Read the input image and split it into tiles
- analyze tile images
  - currently a naive pixel-to-pixel comparision to any tile and calculate the difference
- Use an algorithm to map the tiles using the difference informations
- Generate the image by loading the used tile and putting them into

## Tile Size

The parameters `-j` and `-k` define the size of an tile inside the input image (excluding scale factor).  

`-s` can define a scale factor applied to the generated mosaic.  
Using this is recommended over just upscaling the input image as the analysis can be done in lower resolution and so faster.

## Mapping Algorithms

### Simple

- just maps the best matching tile to any field
- best recreation
- result won't look really random

### AllToAll

- tries to use as many of the tile images as possible

## Tile Comparison

is currently implemented by comparing the source tile against all destination tiles pixel-by-pixel by CIELAB square color distance. This is not very performant, but can allow very precise choices in tiles.

## Memory Consumption

Mosion can comsume up to a few gigabytes of memory.

Tips to cut memory usage

- Use `-s` whenever possible
- As last resort `--no_lab_cache` can be used. will negatively affect performance

## TODO

- [ ] re-implement noise (to increase randomness of tiles)
- [ ] implement any kind of disk caching of comparision data


CHANGELOG.md
# 0.2

- feat: all-to-all mapper
  - improved mapper which tries to match all source tiles
- feat: compose scale factor
  - multiplier for composing image/tile size
- compose: multithreaded image decoding
- optimized memory usage
- reworked mapper to use less vecs
- fixes

src/main.rs
use crate::composing::compose_walls;
use crate::decoding::decode_compare_all;
use crate::meta::Meta;
use crate::decoding::files::collect_files;
use crate::comparer::boring::Boring;
use crate::decoding::split::split_wall;
use std::fs::read;
use std::sync::Arc;
use std::{io::Write, path::PathBuf};
use image::imageops::FilterType;
use crate::util::RefClonable;
use crate::mapping::*;
use image::ImageFormat;

#[macro_use]
pub mod util;

pub mod meta;

pub mod decoding;

pub mod comparer;
pub mod scaler;

pub mod mapping;

pub mod composing;

#[macro_use]
extern crate clap;

fn main() {
    let matches = clap_app!(mosion =>
        (version: "0.2")
        (about: "Mosaic Image Generator")
        (@arg NORECURSE: -nr "No recursing in tile dir")
        (@arg OVERWRITE: -y "Allow overwriting output file")
        (@arg NOLABCACHE: --no_lab_cache "Disable caching of converted wall. Reduces memory usage, severely hurts performance")
        (@arg TILEW: -j +takes_value "Tile Width (default=64)")
        (@arg TILEH: -k +takes_value "Tile Height (default=64)")
        (@arg CSCALE: -s +takes_value "Compose Scale Factor (default=1)")
        (@arg ACHUNKS: -p +takes_value "Analysis concurrency (increases I/O utilization) (default=64)")
        (@arg MAPPER: -m +takes_value "Mapping Algorithm (simple/alltoall) (default=alltoall)")
        
        (@arg INPUT: +required "Input \"Wall\" Image")
        (@arg DIR: +required "Tile Image Dir")
        (@arg OUTPUT: +required "Output Image")
    ).get_matches();

    let recurse = !matches.is_present("NORECURSE");
    let overwrite = !matches.is_present("OVERWRITE");
    let nolabcache = matches.is_present("NOLABCACHE");

    let tw: u32 = matches.value_of("TILEW").unwrap_or("64").parse().expect("Error: non-number at -tw");
    let th: u32 = matches.value_of("TILEH").unwrap_or("64").parse().expect("Error: non-number at -th");
    let cscale: u32 = matches.value_of("CSCALE").unwrap_or("1").parse().expect("Error: non-number at -s");
    let achunks: usize = matches.value_of("ACHUNKS").unwrap_or("64").parse().expect("Error: non-number at -p");

    let input = matches.value_of("INPUT").expect("Error: Input not set");
    let dir = matches.value_of("DIR").expect("Error: Tile Dir not set");
    let output = matches.value_of("OUTPUT").expect("Error: Output not set");

    let (input, dir, output) = (PathBuf::from(input), PathBuf::from(dir), PathBuf::from(output));

    assert!(input.is_file(),"Error: Input is not an existing file");
    assert!(dir.is_dir(),"Error: Tile Dir is not an existing directory");

    if overwrite {
        assert!(!output.is_dir(),"Error: Output is a directory");
    }else{
        assert!(!output.exists(),"Error: Output exists");
    }

    let mapping = matches.value_of("MAPPER").unwrap_or("alltoall");

    assert!(valid_mapping(mapping),"Error: invalid mapper {}",mapping);

    run((tw,th), cscale, recurse, input, dir, output, mapping, nolabcache, achunks);
}

pub fn run(tile_size: (u32,u32), cscale: u32, recurse: bool, input: PathBuf, tile_dir: PathBuf, output: PathBuf, mapping: &str, nolabcache: bool, achunks: usize) {
    println!("Start finding tile images");
    let tile_files = std::thread::spawn(move || {
        collect_files(tile_dir,recurse,false)
    });

    println!("Decode input image");
    //we encode from memory so that image-rs won't guess by extension
    let in_image_0 = read(input).expect("Failed to read input image");
    let in_image_1 = image::load_from_memory(&in_image_0).expect("Failed to decode input image");
    drop(in_image_0);
    let in_image = in_image_1.to_rgba();
    drop(in_image_1);

    println!("Split Image in tiles");

    let (mut walls_parsed,tile_axis) = split_wall::<Boring>(in_image, false, tile_size, FilterType::Triangle, !nolabcache);
    walls_parsed.shrink_to_fit();

    print!("Wait for tile search... ");
    std::io::stdout().flush().unwrap();

    let mut tile_files = tile_files.join().expect("Tile image find thread crashed");
    tile_files.shrink_to_fit();

    println!("DONE");

    if tile_files.is_empty() {panic!("No Tile files");}

    let mut meta = Arc::new(Meta{
        scale: FilterType::Triangle,
        tile_size,
        tile_axis,
        walls_parsed,
        achunks,
    });

    println!("Decode and compare tiles");

    let mut matches = decode_compare_all::<Boring>(tile_files.clone(), meta.refc());

    let meta_ref = Arc::get_mut(&mut meta).unwrap();

    let walls = meta_ref.walls_parsed.len();
    // dealloc parsed walls
    meta_ref.walls_parsed = Vec::new();

    println!("Sort matches");

    matches.sort_unstable_by_key(|t| t.diff );

    println!("Find optimal matches");

    let mut walls_link = vec![None;walls];

    puzzle_with(mapping, matches, &mut walls_link, tile_files.len(),0).unwrap();

    println!("Compose mosaic image");

    let composed = compose_walls::<Boring>(&walls_link, &tile_files, cscale, meta.refc());

    println!("Write composed");

    composed.save_with_format(output, ImageFormat::Png).expect("Failed to write output image");
}